lib/json_api/json_api.rb

```diff
@@ -0,0 +1,70 @@
+require 'sinatra/base'
+require 'sinatra/json'
+require 'sinatra/namespace'
+require 'json_api/game_deserializer'
+require 'json_api/game_serializer'
+require 'json_api/player_deserializer'
+require 'game/game'
+
+module JsonAPI
+  class JsonAPI < Sinatra::Base
+    register Sinatra::Namespace
+
+    def initialize
+      super
+      @player_deserializer = PlayerDeserializer.new
+      @game_serializer = GameSerializer.new
+      @game_deserializer = GameDeserializer.new
+    end
+
+    namespace '/api/v1' do
+      new_game = lambda do
+        begin
+          req_body = request_body
+          players = @player_deserializer.deserialize(req_body["players"])
+          game = Game::Game.new(players, board_size: board_size_or(req_body, 3))
+          json @game_serializer.serialize(game)
+        rescue
+          json "error" => "invalid players provided"
+        end
+      end
+
+      play_game = lambda do
+        begin
+          make_move
+        rescue
+          json "error" => "invalid game or players"
+        end
+      end
+
+      post '/game/new', &new_game
+      post '/game/play', &play_game
+    end
+
+    private
+
+    def make_move
```

Comment:
Feels like this core logic could somehow be pulled out and decoupled from the JSON API? maybe?

Reply:
The deserialiser handling bad input errors makes sense to me. It's like when you try to parse bad JSON and ruby says it's garbage.  

```
      if move != nil && current_player_type == 'human'
        game = game.place_token(move) unless game.game_over?
      elsif current_player_type != 'human'
        game = game.make_move unless game.game_over?
``` is what I was looking at. There's game logic here, like conditionals, that don't seem API specific to me. It's after things are parsed that it happens.

Reply:
@ssunday ah yes, had forgotten about that

Given that the game is a closed loop, it made sense that I use the `place_token` method as apposed to `make_move`, which uses a `ui` interface to get the `HumanPlayer` move. Since this is async, rather than sync like the console client was, I wasn't sure how else to approach it.

Thinking a bit more, I could probably make the controller implement the same `ui` interface and pass this into the game. The only issue is how to get the move from the JSON to the relevant method. I believe that Sinatra creates a new instance of the controller for each request. Therefore, would it be hideous to parse the move from the JSON and store it in an instance variable, which the `get_move` method returns? e.g.

```
# Gets the move from the request and stores it in an instance variable
post '/make_move' do
  ...
  @move = parse_move(json)
  do_game_stuff(self)
  ...
end

# Called by HumanPlayer.
def get_move
  @move
end
```

Reply:
Yeah, that's sort of weird...hmm. How about just passing it through functions? Why do you need to persist anything? Is there anyway you can save the state elsewhere ?

Reply:
I probably still have access to the request, so could just use that. Another option would be to use observer pattern, but that would also require changing the player and console client.

Reply:
If the request has what you need, you can use it

Reply:
Cool. How do you feel abut `begin/rescue/end` in the controller?

Reply:
It looks like the begin/end could be removed too since it's top level, no?

Reply:
I did not know that. Useful, thanks.